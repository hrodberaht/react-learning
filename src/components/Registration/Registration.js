import React, { Component } from 'react';
import { reduxForm, Field } from 'redux-form';
import { withRouter } from 'react-router-dom';
import { PropTypes } from 'prop-types';

const validate = (values) => {
  const errors = {};
  if (!values.login) {
    errors.login = 'Required';
  }
  if (!values.email) {
    errors.email = 'Required';
  }
  if (!values.password) {
    errors.password = 'Required';
  }
  return errors;
};

const asyncValidate = (values) => {
  const { login } = values;
  return fetch('http://localhost:3004/check', {
    method: 'post',
    headers: {
      'Content-Type': 'application/json',
    },
    body: JSON.stringify({ login }),
  })
    .then(data => data.json())
    .then((data) => {
      if (data.message === 'That login is taken') {
        const error = { login: data.message };
        throw error;
      }
    });
};


const renderInput = ({
  input,
  type,
  placeholder,
  meta: {
    touched,
    error,
  },
}) => (
  <p>
    <input {...input} type={type} placeholder={placeholder} />
    {touched && error && <span>{ error }</span>}
  </p>
);


export class Registration extends Component {
redirect = () => {
  this.props.history.push('/login');
}

submit = (values) => {
  const { login, email, password } = values;
  fetch('http://localhost:3004/registration', {
    method: 'post',
    headers: {
      'Content-Type': 'application/json',
    },
    body: JSON.stringify({ login, email, password }),
  })
    .then(data => data.json())
    .then((data) => {
      if (data.message === 'user added') this.redirect();
    })
    .catch(error => console.log(error));
}


render() {
  const { handleSubmit } = this.props;
  return (
    <div>
      <form onSubmit={handleSubmit(this.submit)}>
        <Field
          id="login"
          name="login"
          label="login"
          placeholder="Login"
          type="text"
          component={renderInput}
        />
        <Field
          id="email"
          name="email"
          placeholder="Email"
          type="email"
          component={renderInput}
        />
        <Field
          id="password"
          name="password"
          placeholder="Password"
          type="password"
          component={renderInput}
        />
        <button type="submit" className="btn submit-btn">Submit</button>
      </form>
    </div>
  );
}
}

withRouter(Registration);

export default reduxForm({
  form: 'registration',
  validate,
  asyncValidate,
  asyncChangeFields: ['login'],
})(Registration);

Registration.propTypes = {
  handleSubmit: PropTypes.func.isRequired,
};
