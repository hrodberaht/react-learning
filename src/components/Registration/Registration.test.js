import { Registration, renderInput } from './Registration';

describe('<Registration />', () => {
  const registrationProps = {
    handleSubmit: jest.fn(),
  };
  const renderInputProps = {
    placeholder: 'Login',
    type: 'text',
    meta: {
      touched: true,
      error: 'Required',
    },
  };
  const setupRegistration = buildSetup(Registration, registrationProps, null);
  const setupRenderInput = buildSetup(renderInput, renderInputProps, null);


  it('should <Registration/> render without crash', () => {
    expect(setupRegistration().wrapper).toMatchSnapshot();
  });

  it('should <renderInput/> render without crash', () => {
    expect(setupRenderInput().wrapper).toMatchSnapshot();
  });
});
