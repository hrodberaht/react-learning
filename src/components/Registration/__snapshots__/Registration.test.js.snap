// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Registration /> should <Registration/> render without crash 1`] = `
<div
  className="registration"
>
  <form>
    <Field
      component={[Function]}
      id="login"
      label="login"
      name="login"
      placeholder="Login"
      type="text"
    />
    <Field
      component={[Function]}
      id="email"
      name="email"
      placeholder="Email"
      type="email"
    />
    <Field
      component={[Function]}
      id="password"
      name="password"
      placeholder="Password"
      type="password"
    />
    <button
      className="btn submit-btn"
      type="submit"
    >
      Submit
    </button>
  </form>
</div>
`;

exports[`<Registration /> should <renderInput/> render without crash 1`] = `
<Fragment>
  <input
    className="form-input"
    placeholder="Login"
    type="text"
  />
  <p>
    <span>
      Required
    </span>
  </p>
</Fragment>
`;
